services/workflows-service/src/storage/get-file-storage-manager.ts

```diff
@@ -0,0 +1,81 @@
+import { diskStorage } from 'multer';
+import { getFileName } from '@/storage/get-file-name';
+import { z } from 'zod';
+import multerS3 from 'multer-s3';
+import { S3Client, S3ClientConfig, GetObjectCommand } from '@aws-sdk/client-s3';
+import * as tmp from 'tmp';
+import * as fs from 'fs';
+
+import { Readable } from 'stream';
+export type TLocalFile = string;
+
+export const S3StorageEnvSchema = z.object({
+  AWS_REGION: z.string(),
+  AWS_S3_BUCKET_SECRET: z.string(),
+  AWS_S3_BUCKET_KEY: z.string(),
+});
+
+const generateAwsConfig = (process: NodeJS.ProcessEnv): S3ClientConfig => {
+  const { AWS_REGION, AWS_S3_BUCKET_KEY, AWS_S3_BUCKET_SECRET } = S3StorageEnvSchema.parse(
+    process.env,
+  );
+
+  return {
+    region: AWS_REGION,
+    credentials: {
+      accessKeyId: AWS_S3_BUCKET_KEY,
+      secretAccessKey: AWS_S3_BUCKET_SECRET,
+    },
+  };
+};
+
+const isS3BucketConfigured = (process: NodeJS.ProcessEnv) => {
+  return !!z.string().optional().parse(process.AWS_S3_BUCKET_KEY);
+};
+
+export const fetchDefaultBucketName = (process: NodeJS.ProcessEnv) => {
+  return z.string().parse(process.AWS_S3_BUCKET_NAME);
+};
+
+export const manageFileByProvider = (process: NodeJS.ProcessEnv) => {
+  if (isS3BucketConfigured(process)) {
+    return multerS3({
+      s3: new S3Client(generateAwsConfig(process)),
+      acl: 'private',
+      bucket: fetchDefaultBucketName(process),
+    });
+  } else {
+    return diskStorage({
+      destination: './upload',
+      filename: getFileName,
+    });
+  }
+};
+
+export const downloadFileFromS3 = async (
+  bucketName: string,
+  bucketKey: string,
+): Promise<TLocalFile> => {
+  try {
+    const getObjectCommand = new GetObjectCommand({ Bucket: bucketName, Key: bucketKey });
+    let s3Client = new S3Client(generateAwsConfig(process.env));
+    const response = await s3Client.send(getObjectCommand);
+    const readableStream = response.Body as Readable;
+    const tmpFile = tmp.fileSync();
+    const writableStream = fs.createWriteStream(tmpFile.name);
+
+    return new Promise((resolve, reject) => {
+      readableStream
+        .pipe(writableStream)
+        .on('finish', () => {
+          resolve(tmpFile.name);
+        })
+        .on('error', error => {
+          reject(error);
```

Comment:
critical:
pay attention that u propagate this error to the api, hence to the client, which will return internal s3 errors, which can expose the user to data exposure